Source/DafnyLanguageServer/Language/CachingParser.cs

```diff
@@ -0,0 +1,69 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Security.Cryptography;
+using System.Security.Policy;
+using System.Text;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Dafny.LanguageServer.Language;
+
+public class CachingParser : ProgramParser {
+  private readonly TickingCache<byte[], DfyParseResult> parseCache = new(new HashEquality());
+
+  public CachingParser(ILogger<ProgramParser> logger) : base(logger) {
+  }
+
+  public void Prune() {
+    parseCache.Prune();
+  }
+
+  protected override DfyParseResult ParseFile(DafnyOptions options, TextReader reader, Uri uri) {
+    var (newReader, hash) = ComputeHashFromReader(uri, reader, HashAlgorithm.Create("SHA256")!);
+    if (!parseCache.TryGet(hash, out var result)) {
+      logger.LogDebug($"Parse cache miss for {uri}");
+      result = base.ParseFile(options, newReader, uri);
+      parseCache.Set(hash, result);
+    } else {
+      logger.LogDebug($"Parse cache hit for {uri}");
+    }
+
+    // Clone declarations before returning them, since they are mutable and we don't want to mutate the one in the cache.
+    // We should cache an immutable version of the AST instead: https://github.com/dafny-lang/dafny/issues/4086
+    var cloner = new Cloner(true, false);
+    var clonedResult = result! with {
+      Module = new FileModuleDefinition(cloner, result.Module)
+    };
+    return clonedResult;
+  }
+
+  /// <summary>
+  /// We read the contents of the reader and store them in memory using chunks
+  /// to prevent allocating a large array of memory.
+  /// </summary>
+  private static (TextReader reader, byte[] hash) ComputeHashFromReader(Uri uri, TextReader reader, HashAlgorithm hashAlgorithm) {
+    var result = new List<char[]>();
+    const int chunkSize = 1024;
+    hashAlgorithm.Initialize();
+    var uriBytes = Encoding.UTF8.GetBytes(uri.ToString());
```

Comment:
To be safe I'd suggest adding a delimiter between the URI and the text content, in order to prevent hash collisions where the URI and text content might be "confused":
```suggestion
    // Add NUL delimiter to avoid collisions (otherwise hash("A" + "BC") == hash("AB" + "C"))
    var uriBytes = Encoding.UTF8.GetBytes(uri.ToString() + "\0");
```
(This should be okay on filesystems that disallow NUL bytes in filenames, which is most filesystems AFAIK.)

Such collisions are unlikely to occur naturally, but it's easy to construct a malicious example. Without a delimiter, a file named `foo.dfy//.dfy` with contents `X`, would have the same hash as another file named `foo.dfy` with contents `//.dfyX`.